api/api-pipelines-as-code-internal-v1/src/main/java/com/thoughtworks/go/apiv1/pipelinesascodeinternal/PipelinesAsCodeInternalControllerV1.java

```diff
@@ -85,11 +116,61 @@ public String controllerBasePath() {
     public void setupRoutes() {
         path(controllerBasePath(), () -> {
             before(PREVIEW, this.mimeType, this::setContentType, this::verifyContentType, this.apiAuthenticationHelper::checkAdminUserAnd403);
+            before(CONFIG_FILES, this.mimeType, this::setContentType, this::verifyContentType, this.apiAuthenticationHelper::checkAdminUserAnd403);
 
             post(PREVIEW, this.mimeType, this::preview);
+
+            post(CONFIG_FILES, this.mimeType, this::configFiles);
         });
     }
 
+    String configFiles(Request req, Response res) {
+        File folder = FileUtil.createTempFolder();
+        String responseText;
+        try {
+            JsonReader jsonReader = GsonTransformer.getInstance().jsonReaderFrom(req.body());
+
+            ConfigHelperOptions options = new ConfigHelperOptions(goConfigService.getCurrentConfig(), passwordDeserializer);
+            MaterialConfig materialConfig = MaterialsRepresenter.fromJSON(jsonReader, options);
+
+            if (configRepoService.hasConfigRepoByFingerprint(materialConfig.getFingerprint())) {
+                res.status(409);
+                responseText = MessageJson.create("Material is already being used as a config repository");
+            } else {
+                Map<String, ConfigFileList> pacPluginFiles = new HashMap<>();
+                checkoutFromMaterialConfig(materialConfig, folder);
+                defaultPluginInfoFinder.pluginDisplayNameToPluginId(CONFIG_REPO_EXTENSION).forEach((k, pluginId) -> {
+                    ConfigRepoPlugin crPlugin = (ConfigRepoPlugin) pluginService.partialConfigProviderFor(pluginId);
+                    pacPluginFiles.put(pluginId, crPlugin.getConfigFiles(folder, new ArrayList<>()));
+                });
+                responseText = jsonizeAsTopLevelObject(req, w -> ConfigFileListsRepresenter.toJSON(w, pacPluginFiles));
+            }
+        } catch (Exception ex) {
+            res.status(500);
+            responseText = MessageJson.create(ex.getMessage());
+        } finally {
+            FileUtils.deleteQuietly(folder);
+        }
+
+        return responseText;
+    }
+
+    protected void checkoutFromMaterialConfig(MaterialConfig materialConfig, File folder) throws ExecutionException, InterruptedException {
+        Material material = materialConfigConverter.toMaterial(materialConfig);
+        subprocessExecutionContext.setGitShallowClone(true);
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+        Future future = executor.submit(() -> {
+            List<Modification> modifications = materialService.latestModification(material, folder, subprocessExecutionContext);
+            materialService.checkout(material, folder, Modification.latestRevision(modifications), subprocessExecutionContext);
+        });
+        try {
+            future.get(2l, TimeUnit.MINUTES);
```

Comment:
To be absolutely, 100% correct, I'd suggest that you add a configuration value to SystemEnvironment and define a default of 120 seconds. So that it could be changed by administrator on a specific deployment rather than hard coded here.